add `--why` option to `poetry show`

tgolsson opened the pull request:
# Pull Request Check List

Resolves: #5396

- [x] Added **tests** for changed code.
- [x] Updated **documentation** for changed code. (none found for poetry show)

This adds a `--why` option which when combined with `--tree` and a single package starts the tree at the parent level to show why a specific dependency is included. Optimally this'd walk the tree to the root; but that requires a graph-like structure to easily do/filter, and feels out of scope for this PR. 

# Example output

Without `--why`:
```
$ poetry show requests --tree
requests 2.27.1 Python HTTP for Humans.
├── certifi >=2017.4.17
├── charset-normalizer >=2.0.0,<2.1.0
├── idna >=2.5,<4
└── urllib3 >=1.21.1,<1.27
```

With `--why`:
```
$ poetry show requests --tree --why
cachecontrol 0.12.10 httplib2 caching for requests
└── requests *
    ├── certifi >=2017.4.17
    ├── charset-normalizer >=2.0.0,<2.1.0
    ├── idna >=2.5,<4
    └── urllib3 >=1.21.1,<1.27
requests-toolbelt 0.9.1 A utility belt for advanced users of python-requests
└── requests >=2.0.1,<3.0.0
    ├── certifi >=2017.4.17
    ├── charset-normalizer >=2.0.0,<2.1.0
    ├── idna >=2.5,<4
    └── urllib3 >=1.21.1,<1.27
```

On a package with no parent:

```
$ poetry show --tree cachy --why
Package cachy is a direct dependency.
cachy 0.3.0 Cachy provides a simple yet effective caching library.
```

Forgetting to specify package:
```
$ poetry show --tree --why
Error: --why requires specifying a package name.
```

abn commented:
@tgolsson can you also update doc? 

In general, I like the idea. I'd use it. Would it be feasible to use `--why` without `--tree` as well?

tgolsson commented:
@abn Found the docs in cli.md, is there any other spot? Also added more tests, and added a column with `--why` for lists:

```
$ poetry show --why
attrs              21.4.0    from pytest                                                      Classes Without Boilerplate
cachecontrol       0.12.10                                                                    httplib2 caching for requests
cachy              0.3.0                                                                      Cachy provides a simple yet effective caching library.
certifi            2021.10.8 from requests                                                    Python package for providing Mozilla's CA Bundle.
importlib-metadata 4.11.2    from keyring,pluggy,poetry-core,pre-commit,pytest,tox,virtualenv Read metadata from Python packages
...
```

It won't accept `--why` for a single package still since it includes that info by default:

```
$ poetry show --why requests
Error: --why cannot be used when displaying a single package.
```

It will as before accept `--why` for a tree for a specific package:

```
$ poetry show --why --tree requests
cachecontrol 0.12.10 httplib2 caching for requests
└── requests *
    ├── certifi >=2017.4.17
    ├── charset-normalizer >=2.0.0,<2.1.0
    ├── idna >=2.5,<4
    └── urllib3 >=1.21.1,<1.27
requests-toolbelt 0.9.1 A utility belt for advanced users of python-requests
└── requests >=2.0.1,<3.0.0
    ├── certifi >=2017.4.17
    ├── charset-normalizer >=2.0.0,<2.1.0
    ├── idna >=2.5,<4
    └── urllib3 >=1.21.1,<1.27
```

But not for the global tree listing, since again it contains all info, and `--why` would add nothing.

```$ poetry show --why --tree
Error: --why requires a package when combined with --tree.
```

jacopok commented on src/poetry/console/commands/show.py:
```diff
@@ -65,6 +81,23 @@ def handle(self) -> int | None:
+
+            if not self.option("tree") and package:
+                self.line_error(
+                    "<error>Error: --why cannot be used when displaying a single"
```
Maybe this could be phrased as "--why cannot be used without --tree when displaying a single package"?

jacopok commented:
Thanks for this implementation! `--why` is very idiomatic :+1:

tgolsson commented:
Gentle poke @abn :-)

abn commented:
@tgolsson consider me poked :)

tgolsson commented:
@abn Good ;-) are there any other blockers or steps in order to get this merged? I can rebase it tomorrow.

abn commented on src/poetry/console/commands/show.py:
```diff
@@ -215,9 +276,22 @@ def handle(self) -> int | None:
         write_latest = name_length + version_length + latest_length + 3 <= width
-        write_description = name_length + version_length + latest_length + 24 <= width
+        write_why = self.option("why") and (
+            name_length + version_length + latest_length + required_by_length + 3
```
Maybe put the bits other than 3 in a variable as it's being reused. Nitpick.

tgolsson replied:
Not quite sure I understood you correctly; but maybe done :-)

abn reviewed (commented):
Minor comment. Otherwise looks good to rebase.
